File: 14/14.py
def main():
    sol = dict()
    max_ = 0
    maxcnt = 0
    for i in range(1, 100_0001):
        temp = i
        cnt = 1
        while temp != 1:
            if temp in sol:
                sol[i] = cnt + sol[temp] - 1
                break
            if temp % 2 == 0:
                temp >>= 1
            else:
                temp = 3 * temp + 1
            cnt += 1
        else:
            sol[i] = cnt
        if maxcnt < sol[i]:
            maxcnt = sol[i]
            max_ = i
    return max_


assert main() == 837799

File: 14/14.js
function main() {
  let sol = new Map();
  let maxi = 0,
    maxcnt = 0;

  for (let i = 1; i <= 1000000; i++) {
    let temp = i;
    let cnt = 1;
    let flag = true;
    while (temp !== 1) {
      if (sol.has(temp)) {
        sol.set(i, cnt + sol.get(temp) - 1);
        flag = false;
        break;
      }
      if (temp % 2 === 0) {
        temp /= 2;
      } else {
        temp = temp * 3 + 1;
      }
      cnt++;
    }
    if (flag) {
      sol.set(i, cnt);
    }
    if (maxcnt < sol.get(i)) {
      maxcnt = sol.get(i);
      maxi = i;
    }
  }
  return maxi;
}
console.assert(main(), 837799);
